### PR TITLE
fix(controller): Various v2.12 fixes. Fixes #4798, #4801, #4806

### LINKED ISSUE #4798: v2.12: `pod deleted` + `re-apply` error  = errored workflow
## Summary 

In v2.12, we can get a `pod deleted` error under high load. I believe this is caused by factors interplaying:

1. The workflow completes successfully.
2. A pod is then deleted during clean-up, we'll get a re-queue of the workflow.
3. On the next reconciliation the informer returns the same (and now out of date) workflow as the last reconciliation.
4. The pod has been deleted, so the reconciliation marks the pod as `Error: pod deleted`. The workflow is marked as errored
5. Update fails due to resource version check.
6. Re-apply overwrites the previously successful workflow with an error workflow.
7. If pod GC strategy is on-success, then the TTL controller will error.

Causes:

* `reapplyUpdate` will happily overwrite a completed workflow or node. 
* v2.12 added indexers. I _think_ when any of these errors, the workflow update is lost
* I _think_ that `DEFAULT_REQUEUE_TIME` should be longer, up to 10s.

Solution

* Modify `reapplyUpdate` to check to see it is overwriting a successful workflow or any successful nodes. Error out. This will prevent any future cases of succeeded workflows being marked as error.
* Modify the indexers so that they can never return errors. This will prevent conflict error.
* I don't think that the grace period for recently created pods is needed after these changes. Mark it with at `TODO` to remove.

Relates to #4795, #4634, #4794

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>
Fixes #4798, #4801, #4806
Relates to #3551
Relates to recent comments in #1256

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/indexes/labels.go [hunk 1]
```diff
@@ -18,7 +16,7 @@ func MetaWorkflowPhaseIndexFunc() cache.IndexFunc {
 	return func(obj interface{}) ([]string, error) {
 		v, err := meta.Accessor(obj)
 		if err != nil {
-			return []string{}, fmt.Errorf("object has no meta: %v", err)
```
Comment (alexec) [inline workflow/controller/indexes/labels.go:21]: any error from indexer results in a panic - so I think we should never error

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -646,10 +646,19 @@ func (woc *wfOperationCtx) reapplyUpdate(wfClient v1alpha1.WorkflowInterface, no
 		if err != nil {
 			return nil, err
 		}
+		if currWf.Status.Fulfilled() {
```
Comment (alexec) [inline workflow/controller/operator.go:649]: something about the indexers (no, I'm not sure what) means we are more likely to operate on the list version of the workflow, and potentially error a previously succeeded workflow - this prevents that

File: workflow/executor/pns/pns.go [hunk 3]
```diff
@@ -369,13 +370,19 @@ func (p *PNSExecutor) updateCtrIDMap() {
...
+	Duration: 1 * time.Second,
+	Steps:    7,
+	Factor:   2,
+}
+
 func (p *PNSExecutor) GetTerminatedContainerStatus(containerID string) (*corev1.Pod, *corev1.ContainerStatus, error) {
 	var pod *corev1.Pod
 	var containerStatus *corev1.ContainerStatus
-	err := wait.Poll(1*time.Second, 3*time.Second, func() (bool, error) {
+	err := wait.ExponentialBackoff(backoffOver30s, func() (bool, error) {
```
Comment (alexec) [inline workflow/executor/pns/pns.go:382]: under high load the Kubernetes API can cease to respond within 3s - this mitigates that

File: workflow/ttlcontroller/ttlcontroller.go [hunk 4]
```diff
@@ -137,121 +103,69 @@ func (c *Controller) enqueueWF(obj interface{}) {
...
+	// Additionally, this allows enough time to make sure the double checking that the workflow is actually expired
+	// truly works.
+	addAfter := remaining + time.Second
+	key, _ := cache.MetaNamespaceKeyFunc(obj)
+	log.Infof("Queueing %v workflow %s for delete in %v", wf.Status.Phase, key, addAfter.Truncate(time.Second))
 	c.workqueue.AddAfter(key, addAfter)
 }
 
 func (c *Controller) deleteWorkflow(key string) error {
-	obj, exists, err := c.wfInformer.GetIndexer().GetByKey(key)
```
Comment (alexec) [inline workflow/ttlcontroller/ttlcontroller.go:166]: this is almost a re-write to simplify - if you add a workflow to the TTL queue - we assume it must be deleted

File: workflow/controller/rate_limiters.go [hunk 5]
```diff
@@ -11,7 +11,7 @@ import (
 type fixedItemIntervalRateLimiter struct{}
 
 func (r *fixedItemIntervalRateLimiter) When(interface{}) time.Duration {
-	return env.LookupEnvDurationOr("DEFAULT_REQUEUE_TIME", 2*time.Second)
+	return env.LookupEnvDurationOr("DEFAULT_REQUEUE_TIME", 10*time.Second)
```
Comment (alexec) [inline workflow/controller/rate_limiters.go:14]: this was 10s in v2.11 and I think it ought to be in v2.12 too

**Comment by alexec** [comment]
`TestStopBehavior`

**Comment by alexec** [comment]
`TestStopBehavior`

**Comment by alexec** [comment]
```
stop-terminate : Error Stopped with strategy 'Stop'
    --- FAIL: TestFunctionalSuite/TestStopBehavior (8.12s)
        functional_test.go:585: 
            	Error Trace:	functional_test.go:585
            	            				then.go:54
            	            				then.go:31
            	            				functional_test.go:584
            	Error:      	Not equal: 
            	            	expected: "Failed"
            	            	actual  : "Error"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,2 @@
            	            	-(v1alpha1.NodePhase) (len=6) "Failed"
            	            	+(v1alpha1.NodePhase) (len=5) "Error"
            	            	 
            	Test:       	TestFunctionalSuite/TestStopBehavior
        functional_test.go:586: 
            	Error Trace:	functional_test.go:586
            	            				then.go:54
            	            				then.go:31
            	            				functional_test.go:584
            	Error:      	Should be empty, but was Stopped with strategy 'Stop'
            	Test:       	TestFunctionalSuite/TestStopBehavior
        functional_test.go:589: 
            	Error Trace:	functional_test.go:589
            	            				then.go:54
            	            				then.go:31
            	            				functional_test.go:584
            	Error:      	Not equal: 
            	            	expected: "Failed"
            	            	actual  : "Error"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,2 @@
            	            	-(v1alpha1.NodePhase) (len=6) "Failed"
            	            	+(v1alpha1.NodePhase) (len=5) "Error"
            	            	 
            	Test:       	TestFunctionalSuite/TestStopBehavior
        functional_test.go:590: 
            	Error Trace:	functional_test.go:590
            	            				then.go:54
            	            				then.go:31
            	            				functional_test.go:584
            	Error:      	Not equal: 
            	            	expected: "workflow shutdown with strategy:  Stop"
            	            	actual  : "pod deleted during operation"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-workflow shutdown with strategy:  Stop
            	            	+pod deleted during operation
            	Test:       	TestFunctionalSuite/TestStopBehavior

=== FAIL: test/e2e TestFunctionalSuite (328.72s)
```

**Review by simster7 — APPROVED** [review]

**Pull request merged**